### PR TITLE
Preserve the CBOR representation of scripts

### PR DESCRIPTION
**Pull request by LudvikGalois**
Fixes https://github.com/input-output-hk/cardano-node/issues/4433

### CODE REVIEW

File: bench/tx-generator/src/Cardano/Benchmarking/Script/Types.hs [hunk 1]
```diff
@@ -18,7 +18,7 @@ import           Data.List.NonEmpty
 import           Data.Text (Text)
 
 import           Cardano.Api (AnyCardanoEra, ExecutionUnits, Lovelace, ScriptData, ScriptRedeemer,
-                   TextEnvelope, TxIn)
+                   TextEnvelope, TxIn, WithCBOR)
```
Comment (LudvikGalois) [inline bench/tx-generator/src/Cardano/Benchmarking/Script/Types.hs:21]: That should be `HashableScriptData` - I must have missed fixing it after introducing the `HashableScriptData` type synonym

File: cardano-api/src/Cardano/Api/ScriptData.hs [hunk 2]
```diff
@@ -131,21 +136,30 @@ instance ToCBOR ScriptData where
 instance FromCBOR ScriptData where
   fromCBOR = fromPlutusData <$> decode @Plutus.Data
 
-hashScriptData :: ScriptData -> Hash ScriptData
+type HashableScriptData = WithCBOR ScriptData
+
+pattern AsHashableScriptData :: AsType (WithCBOR ScriptData)
+pattern AsHashableScriptData = AsWithCBOR AsScriptData
```
Comment (LudvikGalois) [inline cardano-api/src/Cardano/Api/ScriptData.hs:142]: I thought a pattern might make it feel more like all the other `AsType`s, but I'm starting to rethink it

File: cardano-api/src/Cardano/Api/ScriptData.hs [hunk 3]
```diff
@@ -131,21 +136,30 @@ instance ToCBOR ScriptData where
...
 
-toAlonzoData :: ScriptData -> Alonzo.Data ledgerera
-toAlonzoData = Alonzo.Data . toPlutusData
+toAlonzoData :: HashableScriptData -> Alonzo.Data ledgerera
+toAlonzoData = Alonzo.binaryDataToData . unsafeMakeBinaryData . getCBORShort
 
-fromAlonzoData :: Alonzo.Data ledgerera -> ScriptData
-fromAlonzoData = fromPlutusData . Alonzo.getPlutusData
+fromAlonzoData :: Alonzo.Data ledgerera -> HashableScriptData
+fromAlonzoData d = case deserialiseFromCBOR (AsWithCBOR AsScriptData) $ Ledger.originalBytes d of
```
Comment (LudvikGalois) [inline cardano-api/src/Cardano/Api/ScriptData.hs:160]:
```suggestion
fromAlonzoData d = case deserialiseFromCBOR AsHashableScriptData $ Ledger.originalBytes d of
```

File: cardano-api/src/Cardano/Api/CBOR.hs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+instance HashAnnotated (WithCBOR a) a StandardCrypto
+
+-- | Create a value of type 'WithCBOR' by first serialising it and then deserialising it.
+-- Note that the value stored here may not be the original value of serialisation doesn't
+-- roundtrip, and this will throw an error if we can't deserialise from the generated
+-- CBOR.
+withCBORViaRoundtrip :: forall a . SerialiseAsCBOR a => a -> WithCBOR a
+withCBORViaRoundtrip x = case deserialiseFromCBOR (AsWithCBOR (proxyToAsType $ Proxy @a)) (serialiseToCBOR x) of
+    Left err -> error $ "withCBORViaRoundtrip: Does not round trip " <> show err
+    Right x' -> x'
```
Comment (LudvikGalois) [inline cardano-api/src/Cardano/Api/CBOR.hs:68]: I'm not happy with this name

File: cardano-api/src/Cardano/Api/ScriptData.hs [hunk 5]
```diff
@@ -131,21 +136,30 @@ instance ToCBOR ScriptData where
 instance FromCBOR ScriptData where
   fromCBOR = fromPlutusData <$> decode @Plutus.Data
 
-hashScriptData :: ScriptData -> Hash ScriptData
+type HashableScriptData = WithCBOR ScriptData
+
+pattern AsHashableScriptData :: AsType (WithCBOR ScriptData)
+pattern AsHashableScriptData = AsWithCBOR AsScriptData
+
+hashScriptData :: HashableScriptData -> Hash ScriptData
```
Comment (LudvikGalois) [inline cardano-api/src/Cardano/Api/ScriptData.hs:144]: I'm not sure if this is needed any more, I think we can just go with `hashAnnotated` from Cardano.Ledger.SafeHash

File: cardano-api/src/Cardano/Api/ScriptData.hs [hunk 6]
```diff
@@ -3,10 +3,17 @@
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE InstanceSigs #-}
```
Comment (LudvikGalois) [inline cardano-api/src/Cardano/Api/ScriptData.hs:6]: I think HLS inserted this for me and it isn't needed. I'll remove it when I next push

**Comment by Jimbo4350** [comment]
So this needs a rebase on `master`. Also in terms of commits, this would be a lot nicer split as follows:
- 1st commit: Diff wrt addition of `Cardano.Api.CBOR` and/or propagation of the new types within `cardano-api`
- 2nd commit: Diff wrt propagation of the new types in cardano-cli
- 3rd commit: Diff wrt `benchmarking` changes.

**Comment by Jimbo4350** [comment]
Subsumed by: #4886

**Pull request closed without merging**